cli/src/commands/parallelize.rs

```diff
@@ -0,0 +1,262 @@
+                command.settings(),
+                &new_parents.into_iter().collect_vec(),
+                &[child],
+                Default::default(),
+            )?;
+        }
+    }
+
+    // Rebase the target commits onto the parents of the root commit.
+    // We already checked that the roots have the same parents, so we can just
+    // use the first one.
+    let target_commits_to_rebase = target_commits
+        .into_iter()
+        // We need to reverse the iterator so that when we rebase the target
+        // commits they will appear in the same relative order in `jj log` that
+        // they were in before being parallelized. The new order is children
+        // before parents.
+        .rev()
+        .collect_vec();
+    rebase_descendants(
+        &mut tx,
+        command.settings(),
+        &new_parents,
+        &target_commits_to_rebase,
+        Default::default(),
+    )?;
+
+    tx.finish(
+        ui,
+        format!("parallelized {} commits", target_commit_ids.len()),
+    )
+}
+
+/// Returns the new parents of the parallelized commits or an error if any of
+/// the following preconditions are not met:
+///
+/// 1. If the heads of the target revset must not have different children.
+/// 2. The roots of the target revset must not have different parents.
+/// 3. The parents of all target revisions except the roots must also be
+///    parallelized. This means that the target revisions must be connected.
+///
+/// The `target_revset` must evaluate to the commits in `target_commits` when
+/// the provided `repo` is used.
+fn check_preconditions_and_get_new_parents(
+    target_revset: &Rc<RevsetExpression>,
+    target_commits: &[Commit],
+    repo: &dyn Repo,
+) -> Result<Vec<Commit>, CommandError> {
+    check_target_heads(target_revset, repo)?;
+    let target_roots = check_target_roots(target_revset, repo)?;
+    check_target_commits_are_connected(&target_roots, target_commits)?;
+
+    // We already verified that the roots have the same parents, so we can just
+    // use the first root.
+    Ok(target_roots[0].parents())
+}
+
+/// Returns an error if the heads of the target revset have children which are
+/// different.
+fn check_target_heads(
+    target_revset: &Rc<RevsetExpression>,
+    repo: &dyn Repo,
+) -> Result<(), CommandError> {
+    let target_heads: Vec<CommitId> = target_revset
+        .heads()
+        .evaluate_programmatic(repo)?
+        .iter()
+        .collect();
+    if target_heads.len() == 1 {
+        return Ok(());
+    }
+    let get_sorted_children = |commit_id: &CommitId| -> Result<_, CommandError> {
+        let expected_children = RevsetExpression::commit(commit_id.clone())
+            .children()
+            .evaluate_programmatic(repo)?
+            .iter()
+            .sorted()
+            .collect_vec();
+        Ok(expected_children)
+    };
+    let expected_children = get_sorted_children(&target_heads[0])?;
+    for head in target_heads[1..].iter() {
```

Comment:
nit: perhaps, this can be optimized to evaluate children of all target_heads once, and check each child's parents. But that wouldn't be important compared to the `.children()` in the main loop.

Maybe we can address that later.

Reply:
Thanks Yuya. I don't know enough to understand how much more efficient this will be, but I've gathered that evaluating the children revset is expensive. I went ahead and implemented your suggestion :)